### DISCUSSION TITLE
How can i add custom agents to the server

### DISCUSSION THREAD

**Discussion by iankemboi22** [roborescue/rcrs-server#70] [General]
i have already developed them (logistics and supply agent) and although i can run it i caanot add it to the simulator to see it

**Comment by gnardin** [comment]
Dear @iankemboi22 
Your question is too general and broad, preventing us to provide a precise answer.
What are you trying to achieve?

RCRS Server supports 3 specific types of agents: Police, Ambulance, and Fire Brigade. When you say you have developed a custom agent, do you mean implemented one of these three types of agents with custom features or a completely new type of agent.

Please provide more information and detail of what you are trying to achieve, so we can support you.

**Comment by iankemboi22** [comment]
I wanted to add two agents to the one already there (Police, Ambulance, and Fire Brigade), i wanted to add Logistics and supply agent and command and control agent but i cannot find documentation or procedures of doing it, i have developed the templates for both but i cannot find where to integrate the to the simulator. the two agents am adding will work synchronously with the rest of the agents that are there how can i achieve this?

**Comment by gnardin** [comment]
Dear @iankemboi22 
There is no documentation on how to add new types of agents in the simulation platform. The documentation details how you can customize the behavior of these three types of agents (Police, Ambulance, Fire Brigade).
The integration of these new types of agents would require some changes in the simulator code itself.

**Comment by iankemboi22** [comment]
Can you provide pointers on what code changes can be made to the simulator
to integrate new types

**Comment by gnardin** [comment]
Dear @iankemboi22 

Find below the list of classes you should be looking at when adding a new type of agent or new commands (i.e., actions) in the environment.

NOTE: The headings starting with lowercase represent the modules or packages in the modules and the headings starting with uppercase represent the classes. Most of the changes required are in the `standards` module, but there are some few things you would need changing in the `gis2` module too.

Please let us know if you have any follow-up questions.

# Rescue Simulation

## standard

### StandardConstants
* Add a new key for the new type

### entities
* Create a new class that `extends Human`
* Create methods representing the actions that this new agent may be able to execute

#### StandardEntityURN
* Add a new constant representing the new type of agent

#### StandardEntityFactory
* Add a new entry in the switch to instantiate the new type of agent using the constant defined in the StandardEntityURN

#### StandardPropertyURN
* Define new properties you may associate with the new type of agent

#### StandardPropertyFactory
* Add the new properties in the switch, so they are instantiate with a specific data type

### kernel

### StandardAgentRegistrar
* Change this class to handle the new type of agents and properties that you want to make visible to others

#### SingleCommandFilter
* Add new commands (actions) created, so they are selectable as a valid command

#### StandardCommandCollector
* Add any new command (action) you have created for the new type of agent in the method `isTriggerCommand`

### kernel.comms

#### ChannelCommunicationModel
#### StandardCommunicationModel
* Alter these classes to allow the new type of agents to be able to communicate and receive communication messages

### messages
* Add new commands that the new type of agents will be able to execute
* The new command class must `extends AbstractCommand`

#### StandardMessageFactory
* Include the new command in the switches

#### StandardMessageURN
* Define the new URN for the new command

### view

#### AnimatedHumanLayer
* If the new type will deal with humans and you want the colors to change depending on the action performed by the new type

#### CommandLayer
* If the new commands 'results' should be represented in the maps with changing colors

#### HumanLayer
* Need to include the new type of agent and specify the initial color and how the color changes over time (like civilians)

## gis2

### GisScenario
* Require for the simulator to be able to take the new type of agents in the scenario XML file

### scenario
* Add new classes to represent the inclusion, removal, etc. of the new type of agents
